sky/provision/kubernetes/instance.py

```diff
@@ -0,0 +1,713 @@
+            '$(prefix_cmd) apt-get update;'
+            '$(prefix_cmd) apt install openssh-server rsync -y; '
+            '$(prefix_cmd) mkdir -p /var/run/sshd; '
+            '$(prefix_cmd) '
+            'sed -i "s/PermitRootLogin prohibit-password/PermitRootLogin yes/" '
+            '/etc/ssh/sshd_config; '
+            '$(prefix_cmd) sed '
+            '"s@session\\s*required\\s*pam_loginuid.so@session optional '
+            'pam_loginuid.so@g" -i /etc/pam.d/sshd; '
+            'cd /etc/ssh/ && $(prefix_cmd) ssh-keygen -A; '
+            '$(prefix_cmd) mkdir -p ~/.ssh; '
+            '$(prefix_cmd) cp /etc/secret-volume/ssh-publickey '
+            '~/.ssh/authorized_keys; '
+            '$(prefix_cmd) chown -R $(whoami) ~/.ssh;'
+            '$(prefix_cmd) chmod 700 ~/.ssh; '
+            '$(prefix_cmd) chmod 600 ~/.ssh/authorized_keys; '
+            '$(prefix_cmd) service ssh restart; '
+            # Eliminate the error
+            # `mesg: ttyname failed: inappropriate ioctl for device`.
+            # See https://www.educative.io/answers/error-mesg-ttyname-failed-inappropriate-ioctl-for-device  # pylint: disable=line-too-long
+            '$(prefix_cmd) sed -i "s/mesg n/tty -s \\&\\& mesg n/" ~/.profile;')
+    ]
+    # TODO(romilb): We need logging and surface errors here.
+    for new_node in new_nodes:
+        run_command_on_pods(new_node.metadata.name, namespace, set_k8s_ssh_cmd)
+
+
+def _label_pod(namespace: str, pod_name: str, label: Dict[str, str]) -> None:
+    """Label a pod."""
+    kubernetes.core_api().patch_namespaced_pod(
+        pod_name,
+        namespace, {'metadata': {
+            'labels': label
+        }},
+        _request_timeout=kubernetes.API_TIMEOUT)
+
+
+def _setup_service(namespace: str, service_spec: Optional[Dict[str, Any]],
+                   pods: List) -> None:
+    if service_spec is not None:
+        logger.info('run_instances: calling create_namespaced_service '
+                    f'(count={len(pods)}).')
+
+        for pod in pods:
+            node_uuid = pod.meta.labels.get(TAG_SVC_POD_UUID)
+            metadata = service_spec.get('metadata', {})
+            metadata['name'] = pod.meta.name
+            service_spec['metadata'] = metadata
+            service_spec['spec']['selector'] = {TAG_SVC_POD_UUID: node_uuid}
+            _ = kubernetes.core_api().create_namespaced_service(
+                namespace, service_spec)
+            # TODO(zhwu): Should we wait for the service to be ready?
+            # Currently, it is fine as we don't create the service by default
+
+
+def run_instances(region: str, cluster_name_on_cloud: str,
+                  config: common.ProvisionConfig) -> common.ProvisionRecord:
+    """Runs instances for the given cluster."""
+    provider_config = config.provider_config
+    namespace = _get_namespace(provider_config)
+    conf = copy.deepcopy(config.node_config)
+    pod_spec = conf.get('pod', conf)
+    service_spec = conf.get('service')
```

Comment:
Are these two required? I think the yaml spec has the pod spec under `config.node_config`, i.e., conf.get('pod') and conf.get('service') always return `None`.

Reply:
It was also adopted from the `node_provider`. Do we have a reason why we added at the first place? I guess it should be fine to remove it?
https://github.com/skypilot-org/skypilot/blob/553082297ff17b100fb0d61c4a70e5fef4d77c2c/sky/skylet/providers/kubernetes/node_provider.py#L385-L386

Reply:
Ahhh its probably a remnant of an old YAML structure I was experimenting with... should be safe to remove I think